File: 864.获取所有钥匙的最短路径.py
#
# @lc app=leetcode.cn id=864 lang=python3
#
# [864] 获取所有钥匙的最短路径
#

# @lc code=start
from curses.ascii import isalpha, islower, isupper
from typing import List


class Solution:
    def shortestPathAllKeys(self, grid: List[str]) -> int:
        keys = set([chr(ord('a') + i) for i in range(26)])
        m, n = len(grid), len(grid[0])
        total_keys = 0
        q = []
        vis = set()
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '@':
                    q = [(i, j, 0, 0)]
                    vis.add((i, j, 0))
                elif grid[i][j].islower():
                    total_keys += 1
        
        while q:
            i, j, step, status = q.pop(0)
            if bin(status).count("1") == total_keys:
                return step
            
            for ni, nj in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)]:
                if not (0 <= ni < m and 0 <= nj < n) or (ni, nj, status) in vis:
                    continue
                if grid[ni][nj] == '#':
                    continue
                elif grid[ni][nj].islower():
                    vis.add((ni, nj, status | (1 << (ord(grid[ni][nj]) - ord('a')))))
                    q.append((ni, nj, step + 1, status | (1 << (ord(grid[ni][nj]) - ord('a')))))
                    continue
                elif grid[ni][nj].isupper() and (1 << (ord(grid[ni][nj]) - ord('A'))) & status == 0:
                    continue
                
                vis.add((ni, nj, status))
                q.append((ni, nj, step + 1, status))

        return -1
# @lc code=end

File: 864.获取所有钥匙的最短路径.cpp
/*
 * @lc app=leetcode.cn id=864 lang=cpp
 *
 * [864] 获取所有钥匙的最短路径
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    using node = tuple<int, int, int>;
    int shortestPathAllKeys(vector<string>& grid) {
        int m = grid.size(), n = grid[0].size();
        int keyCnt = 0;
        vector<vector<vector<bool>>> vis(m, vector<vector<bool>>(n, vector<bool>(1 << keyCnt)));
        deque<tuple<int, int, int, int>> q;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '@') {
                    q.push_back(tuple(i, j, 0, 0));
                    vis[i][j][0] = true;
                } else if (islower(grid[i][j])) {
                    keyCnt += 1;
                }
            }
        }

        int dirs[5] = {1, 0, -1, 0, 1};

        while (!q.empty()) {
            int x, y, step, status;
            tie(x, y, step, status) = q.front();
            q.pop_front();
            int cnt = 0;
            for (int i = 0; i < 6; i++) {
                if ((status >> i) & 1) {
                    cnt += 1;
                }
            }
            if (cnt == keyCnt) {
                return step;
            }
            cout << x << " " << y << endl;

            for (int i = 0; i < 4; i++) {
                int nx = x + dirs[i];
                int ny = y + dirs[i + 1];
                if (!(0 <= nx && nx < m && 0 <= ny && ny < n) || vis[nx][ny][status]) {
                    continue;
                }

                if (grid[nx][ny] == '#') {
                    continue;
                } else if (islower(grid[nx][ny])) {
                    cout << grid[nx][ny] << (1 << (grid[nx][ny] - 'a')) << endl;
                    vis[nx][ny][status | (1 << (grid[nx][ny] - 'a'))] = true;
                    q.push_back(make_tuple(nx, ny, step + 1, status | (1 << (grid[nx][ny] - 'a'))));
                    continue;
                } else if (isupper(grid[nx][ny]) && ((1 << (grid[nx][ny] - 'A')) & status) == 0) {
                    continue;
                }

                vis[nx][ny][status] = true;
                q.push_back(make_tuple(nx, ny, step + 1, status));
            }

        }

        return -1;
    }
};
// @lc code=end

